internal/pkg/agent/cmd/container.go

```diff
@@ -988,3 +991,129 @@ func isContainer(detail component.PlatformDetail) component.PlatformDetail {
 	detail.OS = component.Container
 	return detail
 }
+
+// agentInfo implements the AgentInfo interface, and it used in shouldFleetEnroll.
+type agentInfo struct {
+	id string
+}
+
+func (a *agentInfo) AgentID() string {
+	return a.id
+}
+
+// shouldFleetEnroll returns true if the elastic-agent should enroll to fleet.
+func shouldFleetEnroll(setupCfg setupConfig) (bool, error) {
+	if !setupCfg.Fleet.Enroll {
+		// fleet enrollment is not enabled
+		return false, nil
+	}
+
+	if setupCfg.Fleet.Force {
+		// fleet enrollment is enforced
+		return true, nil
+	}
+
+	agentCfgFilePath := paths.AgentConfigFile()
+	_, err := os.Stat(agentCfgFilePath)
+	if os.IsNotExist(err) {
+		// config file does not exist so elastic-agent should enroll
+		return true, nil
+	}
+
+	// read the stored elastic-agent state to extract the previous fleet url and enrollment token
+	ctx := context.Background()
+	store, err := storage.NewEncryptedDiskStore(ctx, agentCfgFilePath)
+	if err != nil {
+		return false, fmt.Errorf("failed to instantiate encrypted disk store: %w", err)
+	}
+
+	reader, err := store.Load()
+	if err != nil {
+		return false, fmt.Errorf("failed to load from disk store: %w", err)
+	}
+
+	cfg, err := config.NewConfigFrom(reader)
+	if err != nil {
+		return false, fmt.Errorf("failed to read from disk store: %w", err)
+	}
+
+	storedConfig, err := configuration.NewFromConfig(cfg)
+	if err != nil {
+		return false, fmt.Errorf("failed to read from disk store: %w", err)
+	}
+
+	storedFleetHosts := storedConfig.Fleet.Client.GetHosts()
+	if len(storedFleetHosts) == 0 || !slices.Contains(storedFleetHosts, setupCfg.Fleet.URL) {
+		// fleet url changed thus elastic-agent should enroll
+		return true, nil
+	}
+
+	// here we don't return true if there is no previously saved enrollment token hash
+	// as an older version of elastic-agent won't have one.
+	if len(storedConfig.Fleet.EnrollmentTokenHash) > 0 {
+		// we have previously saved an enrollment token hash
+		enrollmentHashBytes, err := base64.StdEncoding.DecodeString(storedConfig.Fleet.EnrollmentTokenHash)
+		if err != nil {
+			return false, fmt.Errorf("failed to decode hash: %w", err)
+		}
+
+		err = crypto.ComparePBKDF2HashAndPassword(enrollmentHashBytes, []byte(setupCfg.Fleet.EnrollmentToken))
+		switch {
+		case errors.Is(err, crypto.ErrMismatchedHashAndPassword):
+			// the enrollment token hash does not match thus elastic-agent should enroll
+			return true, nil
+		case err != nil:
+			return false, fmt.Errorf("failed to compare hash: %w", err)
+		}
+	}
+
+	// Although fleet url and enrollment token are the same, we must also validate the api token.
+	// The agent might have been unenrolled from kibana which doesn't require for fleet url and
+	// enrollment token need to be changed
```

Comment:
Who unenrolled it? Like it was a manual action in the UI to unenroll it? Should it then auto re-enroll?

Reply:
ok I hear you, so I go I unenroll agent from Kibana and then I restart the pods, what would you expect the agent to do? Personally I think that since I give it a valid fleet url and a token and this is unenrolled, it should re-enroll. Also this user [here](https://github.com/elastic/elastic-agent/issues/3586#issuecomment-2602306451) 🙂

Reply:
Yeah that does make sense. Could you maybe add more of that context to the comment?

Reply:
I update the comment, feel free to unresolve if you believe that more context is needed 🙂